### PR TITLE
Jetty 12.1.x 12088 core requested session ID source

### LINKED ISSUE #12088: No jetty-core API to determine if sessionId is obtained from cookie or URL
**Jetty version(s)**
12.0.x

**Jetty Environment**
core

**Description**
In the servlet API there is the methods `isRequestedSessionIdFromCookie()` and `isRequestedSessionIdFromURL()` on `HttpServletRequest`. 

However it does not seem possible to determine this from jetty-core.

### PR DESCRIPTION
**Pull request by gregw**
Fixes #12088 by adding a mechanism to expose the `RequestedSession` record and the source of the ID.
Replaces #12138

### CODE REVIEW

File: jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java [hunk 1]
```diff
@@ -1379,12 +1362,89 @@ private void shutdownSessions()
...
+        public static final String ATTRIBUTE = "org.eclipse.jetty.session.RequestedSession";
+        public static final String ID_FROM_COOKIE = "cookie";
+        public static final String ID_FROM_URI_PARAMETER = "uri";
+
+        /**
+         * Get the {@code RequestedSession} by attribute
+         * @param request The attributes to query
+         * @return The found {@code RequestedSession} or {@link #NO_REQUESTED_SESSION} if none found. Never {@code null}.
+         */
+        public static RequestedSession byAttribute(Attributes request)
```
Comment (janbartel) [inline jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java:1388]: If we have the ability to retrieve the information via individual attributes, why do we need to expose the `RequestedSession` record at all? I'm not really sure it should be exposed, it was only ever intended as an information carrier for the internal implementation. Also, shouldn't the attributes have constant definitions?

**Review by janbartel — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java:1388]: It was already exposed to some extent to anything extending AbstractSessionManager.  It is used by both core and servlet session handlers.

**Review by janbartel — APPROVED** [review]

**Review by lachlan-roberts — APPROVED** [review]

**Comment by lachlan-roberts** [comment]
@gregw seems to be some test failures on this PR

**Comment by gregw** [comment]
The CI failures of `org.eclipse.jetty.tests.distribution.DistributionTests#testWebAppWithProxyAndJPMS` appear to also be in jetty-12.1.x HEAD, so I'm inclined to merge this.

**Pull request merged**